akka-persistence/src/main/scala/akka/persistence/PersistentActor.scala

```diff
@@ -99,6 +99,58 @@ object Recovery {
 }
 
 /**
+ * This defines how to handle the current received message which failed to stash, when the size of
+ * Stash exceeding the capacity of Stash.
+ */
+sealed trait StashOverflowStrategy
+
+/**
+ * Discard the message to [[DeadLetter]].
+ */
+case object DiscardToDeadLetterStrategy extends StashOverflowStrategy {
+  /**
+   * Java API: get the singleton instance
+   */
+  def getInstance = this
+}
+
+/**
+ * Throw [[StashOverflowException]], hence the persistent actor will starting recovery 
+ * if guarded by default supervisor strategy. 
+ * Be carefully if used together with persist/persistAll or has many messages needed 
+ * to replay.
+ */
+case object ThrowOverflowExceptionStrategy extends StashOverflowStrategy {
+  /**
+   * Java API: get the singleton instance
+   */
+  def getInstance = this
+}
+
+/**
+ * Reply to sender with predefined response, and discard the received message silently.
+ * @param response the message replying to sender with
+ */
+final case class ReplyToStrategy(response: Any) extends StashOverflowStrategy
+
+/**
+ * Implement this interface in order to configure the stashOverflowStrategy for
+ * the internal stash of persistent actor.
+ * An instance of this class must be instantiable using a no-arg constructor.
+ */
+trait StashOverflowStrategyConfigurator {
```

Comment:
We don't want these to be extendable, but I see that you need it to be able to support the `ReplyToStrategy` with a custom message. I'm fine with this.

```
cfg.getString("internal-stash-overflow-strategy" match {
  case "overflow-exception" => ThrowOverflowExceptionStrategy
  case "